shared/modules/Storage/ExtensionStore.ts

```diff
@@ -0,0 +1,201 @@
+    }
+    // we use this flag to avoid flooding sentry with a ton of errors:
+    // once data persistence fails once and it flips true we don't send further
+    // data persistence errors to sentry
+    this.dataPersistenceFailing = false;
+    this.mostRecentRetrievedState = null;
+    this.migrator = migrator;
+  }
+
+  /**
+   * Persists the MetaMask state tree on the 'data' key of the
+   * browser.storage.local API. This function will first do some sanity checks
+   * to determine if it is likely for the operation to succeed. It will throw
+   * an error in the following cases.
+   * 1. The browser does not support the browser.storage.local API.
+   * 2. No state object was provided to the function. As more of the codebase
+   * is migrated to TypeScript, this should become less of a possibility but
+   * never impossible.
+   * 3. The metadata property is not set on the class which is required before
+   * setting state. This ensures the 'meta' key of the storage is set.
+   *
+   * @param state - The state to persist to the data key of the local store
+   * @returns void
+   */
+  async set(state: IntermediaryStateType) {
+    if (!this.isSupported) {
+      throw new Error(
+        'Metamask- cannot persist state to local store as this browser does not support this action',
+      );
+    }
+    if (!state) {
+      throw new Error('MetaMask - updated state is missing');
+    }
+    if (!this.metadata) {
+      throw new Error(
+        'MetaMask - metadata must be set on instance of ExtensionStore before calling "set"',
+      );
+    }
+    try {
+      // we format the data for storage as an object with the "data" key for the controller state object
+      // and the "meta" key for a metadata object containing a version number that tracks how the data shape
+      // has changed using migrations to adapt to backwards incompatible changes
+      await this.#set({ data: state, meta: this.metadata });
+      if (this.dataPersistenceFailing) {
+        this.dataPersistenceFailing = false;
+      }
+    } catch (err) {
+      if (!this.dataPersistenceFailing) {
+        this.dataPersistenceFailing = true;
+        captureException(err);
+      }
+      log.error('error setting state in local store:', err);
+    }
+  }
+
+  /**
+   * Returns all of the keys currently saved
+   */
+  async get() {
+    /**
+     * If chrome.storage.local is not available, return the default state tree
+     * which will not be persisted. This should probably be a bug that we
+     * report to sentry.
+     *
+     * TODO: Investigate what happens in this case and log sentry report.
+     */
+    if (!this.isSupported) {
+      return this.generateFirstTimeState();
+    }
+    try {
+      const result = await this.#get();
+      // extension.storage.local always returns an obj
+      // if the object is empty, treat it as undefined
+      if (isEmpty(result)) {
+        this.mostRecentRetrievedState = null;
+        this.stateCorruptionDetected = true;
```

Comment:
I might be missing it, but I don't see `this.stateCorruptionDetected` being used anywhere. I see it being set to `true` and `false`, but then not used otherwise

Reply:
Now used in unit tests as of 70e61917c9

Will be used in application code in a following PR